[Transform] [Quantization] Add QuTLASS support to vLLM

LopezCastroRoberto opened the pull request:
## Purpose

This pull request brings in the QuTLASS library: https://github.com/iST-DASLab/qutlass

QuTLASS is a high-performance library designed for low-precision kernel support in deep learning quantization, built on top of [NVIDIA CUTLASS](https://github.com/NVIDIA/cutlass).

QuTLASS v0.1.0 introduces 4-bit microscaling routines tailored for Large Language Model (LLM) inference on NVIDIA Blackwell GPUs.

* Online rotations:
   - Fused transform + quantization + scale computation.
      - Rotation matrices loaded at runtime, allowing any transformation to be applied. 
   - Support for both NVFP4 and MXFP4 microscaling formats.
   - Multiple rotation sizes (16/32/64/128).
* MXFP4 matmul kernel support powered by CUTLASS.
  - QuTLASS is compatible with any matmul backend supporting microscaling formats (e.g., CUTLASS, FlashInfer).
* Multiple quantization schemes:
   - [Quartet](https://arxiv.org/pdf/2505.14669) (i.e., [Quest-like](https://arxiv.org/abs/2502.05003)).
   - Abs-Max.

## Microbenchmarking

* ```benchmarks/kernels/bench_mxfp4_qutlass.py```
* ```benchmarks/kernels/bench_nvfp4_qutlass.py```

<p align="center">
  <img src="https://raw.githubusercontent.com/IST-DASLab/qutlass/830d2c4537c7396e14a02a46fbddd18b5d107c65/assets/flops.svg" alt="Llama-32B MXFP4:MXFP4" width="45%">
  <img src="https://raw.githubusercontent.com/IST-DASLab/qutlass/830d2c4537c7396e14a02a46fbddd18b5d107c65/assets/flops_nvfp4.svg" alt="Llama-32B NVFP4:NVFP4" width="45%">
  <br>
  <em>QuTLASS performance on a single Qwen3-32B layer with NVIDIA RTX5090 GPU</em>
</p>

<p align="center">
  <img src="https://raw.githubusercontent.com/IST-DASLab/qutlass/830d2c4537c7396e14a02a46fbddd18b5d107c65/assets/flops_sm100.svg" alt="Llama-70B MXFP4:MXFP4" width="45%">
  <img src="https://raw.githubusercontent.com/IST-DASLab/qutlass/830d2c4537c7396e14a02a46fbddd18b5d107c65/assets/flops_sm100_nvfp4.svg" alt="Llama-70B NVFP4:NVFP4" width="45%">
  <br>
  <em>QuTLASS performance on a single Llama-3.1-70B layer with NVIDIA B200 GPU</em>
</p>

## [WIP] End-to-end
* ```python benchmarks/benchmark_latency.py```
    * ```daslab-testing/Llama-3.3-70B-Instruct-FPQuant-GPTQ-MXFP4-hadamard```
    * ```meta-llama/Llama-3.3-70B-Instruct```
<p align="center">
<img width="60%" height="600" alt="vLLM_b200" src="https://github.com/user-attachments/assets/fe8c33b9-7118-41fd-80ae-fa65d5ae960f" />
</p>

### FP16
| Quantization | Llama: MMLU-CoT | GSM8k | Hellaswag | Winogrande | **Average** | **Recovery %** |
|--------------|----------------|-------|-----------|------------|-------------|----------------|
| N/A          | **0.866**      | **0.951** | **0.862**   | **0.849**  | **0.882**   | -              |

### MXFP4
| Quantization        | Llama: MMLU-CoT | GSM8k | Hellaswag | Winogrande | **Average** | **Recovery %** |
|---------------------|----------------|-------|-----------|------------|-------------|----------------|
| RTN                 | 0.834          | 0.927 | 0.839     | 0.815      | 0.854       | 96.8           |
| RTN + HAD (GS32)    | 0.839          | 0.936 | 0.841     | 0.836      | 0.863       | 97.8           |
| GPTQ                | 0.838          | **0.945** | 0.844     | 0.826      | 0.863       | 97.9           |
| GPTQ + HAD (GS32)   | 0.848          | **0.945** | 0.847     | 0.831      | 0.868       | **98.4**       |



## Testing

* ```tests/kernels/quantization/test_mxfp4_qutlass.py```
* ```tests/kernels/quantization/test_nvfp4_qutlass.py```

voipmonitor commented:
@LopezCastroRoberto does this PR support gpt-oss on sm120 ? How to exactly test some mxfp4 models with this PR? Would love to test rtx 6000 pro on this

LopezCastroRoberto commented on vllm/_custom_ops.py:
```diff
@@ -2012,3 +2012,57 @@ def onednn_scaled_mm(
+
+def fusedQuantizeMx(a: torch.Tensor,
+                    b: torch.Tensor,
+                    *,
```
Means all arguments that come after the * must be passed by keyword, not by position. My point was to make the API clearer and less error-prone.

LopezCastroRoberto commented on benchmarks/kernels/bench_nvfp4_qutlass.py:
```diff
@@ -0,0 +1,160 @@
+    # Quantize activation on-the-fly
+    def run():
+        input_hf_e2m1, input_hf_e8m0 = fusedQuantizeNv(a, forward_hadamard_matrix, global_scale)
+        input_hf_scale_block = to_blocked(input_hf_e8m0, True).view(-1,K//16)
```
yes—the very first time is slower, but after that it's cached

LopezCastroRoberto commented on vllm/qutlass_utils/utils.py:
```diff
@@ -0,0 +1,173 @@
+    return (a + b - 1) // b
+
+
+def to_blocked(input_matrix, use_triton_kernel: bool = False) -> Tensor:
```
How about keeping one ```to_blocked``` but making the backend explicit ```(e.g. backend="torch" | "triton" | "auto")```?

LopezCastroRoberto committed:
fixing merge issues with main branch

Signed-off-by: LopezCastroRoberto <roberto.lopez.castro@udc.es>

LopezCastroRoberto committed:
fix: apply pre-commit formatting

Signed-off-by: LopezCastroRoberto <roberto.lopez.castro@udc.es>

LopezCastroRoberto committed:
add spdx license

Signed-off-by: LopezCastroRoberto <roberto.lopez.castro@udc.es>

LopezCastroRoberto committed:
fix: apply pre-commit formatting

Signed-off-by: LopezCastroRoberto <roberto.lopez.castro@udc.es>

LopezCastroRoberto committed:
style: apply auto-fixes from pre-commit

Signed-off-by: LopezCastroRoberto <roberto.lopez.castro@udc.es>

kylesayrs reviewed (approved):
LGTM!

LopezCastroRoberto committed:
Merge branch 'main' into transforms

Signed-off-by: Roberto L. Castro <38211239+LopezCastroRoberto@users.noreply.github.com>

LopezCastroRoberto committed:
style: apply auto-fixes from pre-commit

Signed-off-by: LopezCastroRoberto <roberto.lopez.castro@udc.es>

mgoin commented on cmake/external_projects/qutlass.cmake:
Does this require some minimum CUDA version?

BlackSamorez committed:
Basic FP-Quant integration

Signed-off-by: Andrei Panferov <andrei@panferov.org>

minor fixes

eager works

eager tests

custom ops fake fix

eager works

eager tests

removed extra op

style

Signed-off-by: Andrei Panferov <andrei@panferov.org>

BlackSamorez committed:
custom ops fake fix

Signed-off-by: Andrei Panferov <andrei@panferov.org>

BlackSamorez commented:
Fixed `register_fake`s. They had a wrong namespace (`_C::_qutlass_C` instead of just `"_qutlass_C`) and quite a few kernels (`matmul_mxf4_bf16_tn`, `matmul_ada_mxf4_bf16_tn`, `fused_quantize_nv`) didn't have fake impls at all.

LopezCastroRoberto committed:
use pytest and cuda version check

Signed-off-by: LopezCastroRoberto <roberto.lopez.castro@udc.es>

LopezCastroRoberto committed:
adding hint to cmake for CI

Signed-off-by: LopezCastroRoberto <roberto.lopez.castro@udc.es>

LopezCastroRoberto committed:
Merge branch 'vllm-project:main' into transforms

LopezCastroRoberto committed:
re-define fake methods

Signed-off-by: LopezCastroRoberto <roberto.lopez.castro@udc.es>

LopezCastroRoberto committed:
style: apply auto-fixes from pre-commit

Signed-off-by: LopezCastroRoberto <roberto.lopez.castro@udc.es>

LopezCastroRoberto committed:
[CI/Build] limit to sm100

Signed-off-by: LopezCastroRoberto <roberto.lopez.castro@udc.es>

LopezCastroRoberto committed:
[CI/Build] flip sm order

Signed-off-by: LopezCastroRoberto <roberto.lopez.castro@udc.es>

LopezCastroRoberto committed:
Merge pull request #1 from LopezCastroRoberto/integration

Integration

LopezCastroRoberto committed:
fix FP-Quant integration

Signed-off-by: LopezCastroRoberto <roberto.lopez.castro@udc.es>

LopezCastroRoberto committed:
Merge branch 'main' into transforms

Signed-off-by: Roberto L. Castro <38211239+LopezCastroRoberto@users.noreply.github.com>

LopezCastroRoberto committed:
style: apply auto-fixes from pre-commit

Signed-off-by: LopezCastroRoberto <roberto.lopez.castro@udc.es>

mgoin reviewed (changes requested):
Requesting changes for qutlass_utils

LopezCastroRoberto committed:
change dir qutlass utils

Signed-off-by: LopezCastroRoberto <roberto.lopez.castro@udc.es>

LopezCastroRoberto committed:
remove noqa e501

Signed-off-by: LopezCastroRoberto <roberto.lopez.castro@udc.es>

LopezCastroRoberto committed:
Merge branch 'vllm-project:main' into transforms

LopezCastroRoberto committed:
isolate QuTLASS compile flags

Signed-off-by: LopezCastroRoberto <roberto.lopez.castro@udc.es>

LopezCastroRoberto committed:
fix docstring

Signed-off-by: LopezCastroRoberto <roberto.lopez.castro@udc.es>

mgoin committed:
Merge branch 'main' into transforms

bbrowning commented on cmake/external_projects/qutlass.cmake:
```diff
@@ -0,0 +1,97 @@
+  if(QUTLASS_ARCHS MATCHES "10\\.0a")
+    set(QUTLASS_TARGET_CC 100)
+  elseif(QUTLASS_ARCHS MATCHES "12\\.0a")
+    set(QUTLASS_TARGET_CC 120)
```
This excludes the DGX Spark (SM121) from using QuTLASS, since it is not `12.0a`. If you adjust this to allow `12.0f` or `12.1a`, this will also get built for the spark. I built this on my Spark locally and all tests in `test_mxfp4_qutlass.py` and `test_nvfp4_qutlass.py` pass, but I'm not sure if that's all that's needed to verify this would work for that hardware.

Is there a way I can verify this is getting used in a running vLLM, outside of those tests?

bbrowning commented:
> There have been [reports](https://github.com/IST-DASLab/FP-Quant/issues/15) of NVFP4 QuTLASS failing on RTX 5090 and DGX Spark too.

I tested locally loosening up the cmake CUDA_ARCHS matching  and was able to run an example FP-Quant model fine on a DGX Spark building vLLM from source. I don't have an RTX 5090 to test, but on the surface I would have assumed it already matched this CUDA_ARCHS check.

BlackSamorez commented:
If I remember correctly, the kernel dispatch is static w.r.t. arch because the kernels were separately tuned for each architecture. Compiling for a non-supported arch would throw runtime errors. @LopezCastroRoberto would we have to re-tune the kernels for `12.1a`?
